com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ServiceRemoval.java

```diff
@@ -0,0 +1,138 @@
+            client.unbindService(application, serviceInstance.getService());
+        }
+    }
+
+    private void logBindings(StepLogger stepLogger, List<CloudServiceBinding> bindings) {
+        stepLogger.debug(Messages.SERVICE_BINDINGS_EXISTS, secureSerializer.toJson(bindings));
+    }
+
+    private void deleteServiceKeys(CloudControllerClient client, StepLogger stepLogger, List<CloudServiceKey> serviceKeys) {
+        for (CloudServiceKey serviceKey : serviceKeys) {
+            stepLogger.info(Messages.DELETING_SERVICE_KEY_FOR_SERVICE, serviceKey.getName(), serviceKeys);
+            client.deleteServiceKey(serviceKey);
+        }
+    }
+
+    private void deleteService(CloudControllerClient client, StepLogger stepLogger, ServiceInstanceWithServiceKeys service) {
+        stepLogger.info(Messages.DELETING_SERVICE, service.getServiceInstance()
+                                                          .getName());
+        client.deleteService(service.getServiceInstance()
+                                    .getService());
+        stepLogger.debug(Messages.SERVICE_DELETED, service.getServiceInstance()
+                                                          .getName());
+    }
+
+    private void processException(ExecutionWrapper execution, Exception e, CloudServiceInstance serviceInstance) {
+        if (e instanceof CloudOperationException) {
+            e = evaluateCloudOperationException(execution, (CloudOperationException) e, serviceInstance.getName(),
+                                                serviceInstance.getService()
+                                                               .getLabel());
+            if (e == null) {
+                return;
+            }
+        }
+        wrapAndThrowException(e, serviceInstance);
+    }
+
+    private CloudOperationException evaluateCloudOperationException(ExecutionWrapper execution, CloudOperationException e,
+                                                                    String serviceName, String label) {
+        if (e.getStatusCode() == HttpStatus.NOT_FOUND) {
+            execution.getStepLogger()
+                     .warn(MessageFormat.format(Messages.COULD_NOT_DELETE_SERVICE, serviceName), e,
+                           ExceptionMessageTailMapper.map(configuration, CloudComponents.SERVICE_BROKERS, label));
+            return null;
+        }
+        if (e.getStatusCode() == HttpStatus.BAD_GATEWAY) {
+            StepsUtil.setServiceOffering(execution.getContext(), Constants.VAR_SERVICE_OFFERING, label);
+            return new CloudServiceBrokerException(e);
+        }
+        return new CloudControllerException(e);
+
+    }
+
+    private void wrapAndThrowException(Exception e, CloudServiceInstance serviceInstance) {
+        String msg = buildNewExceptionMessage(e, serviceInstance.getService());
+        throw new SLException(e, msg);
+    }
+
+    private String buildNewExceptionMessage(Exception e, CloudService service) {
```

Comment:
The method name is not very descriptive. Reading it, does not point at what exception message exactly it builds, especially the 'New' keyword. I had to check in the implementation to understand that it builds an error message wen deleting a service.